nscplugin/src/main/scala-next/scala/scalanative/nscplugin/NirGenExpr.scala

```diff
@@ -0,0 +1,2642 @@
+import scala.scalanative.nir
+import nir._
+import scala.scalanative.util.ScopedVar.scoped
+import scala.scalanative.util.unsupported
+import scala.scalanative.util.StringUtils
+import dotty.tools.dotc.ast.desugar
+import dotty.tools.dotc.util.Property
+
+trait NirGenExpr(using Context) {
+  self: NirCodeGen =>
+  import positionsConversions.fromSpan
+
+  sealed case class ValTree(value: nir.Val) extends Tree
+  sealed case class ContTree(f: () => nir.Val) extends Tree
+
+  class ExprBuffer(using fresh: Fresh) extends FixupBuffer {
+    buf =>
+    def genExpr(tree: Tree): Val = {
+      tree match {
+        case EmptyTree      => Val.Unit
+        case ValTree(value) => value
+        case ContTree(f)    => f()
+        case tree: Apply =>
+          val updatedTree = lazyValsAdapter.transformApply(tree)
+          genApply(updatedTree)
+        case tree: Assign         => genAssign(tree)
+        case tree: Block          => genBlock(tree)
+        case tree: Closure        => genClosure(tree)
+        case tree: Labeled        => genLabelDef(tree)
+        case tree: Ident          => genIdent(tree)
+        case tree: If             => genIf(tree)
+        case tree: JavaSeqLiteral => genJavaSeqLiteral(tree)
+        case tree: Literal        => genLiteral(tree)
+        case tree: Match          => genMatch(tree)
+        case tree: Return         => genReturn(tree)
+        case tree: Select         => genSelect(tree)
+        case tree: This           => genThis(tree)
+        case tree: Try            => genTry(tree)
+        case tree: Typed          => genTyped(tree)
+        case tree: TypeApply      => genTypeApply(tree)
+        case tree: ValDef         => genValDef(tree)
+        case tree: WhileDo        => genWhileDo(tree)
+        case _ =>
+          throw FatalError(
+            s"""Unexpected tree in genExpr: `${tree}`
+             raw=$tree
+             pos=${tree.span}
+             """
+          )
+      }
+    }
+
+    def defnNir(using Context): NirDefinitions = NirDefinitions.get
+
+    lazy val SafeZoneHandle = new Property.Key[Val]
+
+    private def safeZoneHandle(tree: Tree): Val =
+      tree.getAttachment(SafeZoneHandle).getOrElse(Val.Null)
+
+    def genApply(app: Apply): Val = {
+      app match {
+        case Apply(_, List(sz, tree))
+            if app.fun.symbol == defnNir.SafeZoneCompat_withSafeZone =>
+          // For new expression with a specified safe zone, e.g. `new {sz} T(...)`,
+          // it's translated to `withSafeZone(sz, new T(...))` in TyperPhase.
+          tree match {
+            case Apply(Select(New(_), nme.CONSTRUCTOR), _) =>
+            case Apply(fun, _) if fun.symbol == defn.newArrayMethod =>
+            case _ => 
+              report.error(s"Unexpected tree in withSafeZone: `${tree}`", tree.srcPos)
+          }
+          // Extract the handle of `sz` and put it into the attachment of `new T(...)`.
+          val handle = genExpr(Select(sz, termName("handle")))
```

Comment:
isn't this a violation of the access modifiers on `handle` when someone tries to allocate outside of the `scalanative` package?

Reply:
Hmm, yes and no. If someone calls `sz.handle` outside the `scalanative` package, he will get a ReferenceError saying the handle method is not accessible, which is generated in typer phase by the typer (the type assigner, more exactly. I might be wrong since I'm not familiar enough with the typer..). For this internal function `withSafeZone` (whose name is now updated to `allocate`), the call to `sz.handle` is directly manipulated as a typed tree and skips the typer phase, so it won't trigger the ReferenceError.

Reply:
If it works that way currently then that's cool, I still wonder about potential future proofing if a future scala native version implements strict access checking at the linking phase e.g.